### ISSUE TITLE
[DDC] Uncaught TypeError for Web app with universal_io dependency in debug mode

### ISSUE THREAD

**Issue by creativecreatorormaybenot** [flutter/flutter#66859]
# Bug

**Note** that this is a ***regression** from *beta `1.22.0-12.1.pre` (previous)* to *beta `1.22.0-12.2.pre` (latest)*. I know this because I opened https://github.com/flutter/flutter/issues/64011, which was fixed with https://github.com/dart-lang/sdk/issues/43453 - however, the same version transition introduced this new bug. The `universal_io` dependecy did work with DDC before.

## Steps to Reproduce

0. Use channel `beta` or later with `flutter config --enable-web` (tested on `1.22.0-12.2.pre • channel beta`)
1. Run `flutter create bug`.
2. Include a `universal_io: ^1.0.1` dependency and `import 'package:universal_io/io.dart';` in `main.dart`.
3. **Include a `File mock() => File('mock');` top-level function for the tree shaker** (note that `File` comes from `universal_io` in this case).
4. Run `flutter run -d chrome`.
5. Open the JavaScript console after Chrome has launched.
6. See the following error and a blank page:

<details>
  <summary>Uncaught TypeError: Class extends value function () { [native code] } is not a constructor or null</summary>

```
Uncaught TypeError: Class extends value function () { [native code] } is not a constructor or null
    at http._ServiceObject.<computed> (:4200/packages/universal_io/src/io_impl_js/sync_socket.dart.lib.js:11145)
    at Object.applyMixin (:4200/dart_sdk.js:3880)
    at load__packages__universal_io__src__io_impl_js__sync_socket_dart (:4200/packages/universal_io/src/io_impl_js/sync_socket.dart.lib.js:13903)
    at Object.execCb (:4200/require.js:1696)
    at Module.check (:4200/require.js:883)
    at Module.<anonymous> (:4200/require.js:1139)
    at :4200/require.js:134
    at :4200/require.js:1189
    at each (:4200/require.js:59)
    at Module.emit (:4200/require.js:1188)
    at Module.check (:4200/require.js:938)
    at Module.enable (:4200/require.js:1176)
    at Module.init (:4200/require.js:788)
    at callGetModule (:4200/require.js:1203)
    at Object.completeLoad (:4200/require.js:1590)
    at HTMLScriptElement.onScriptLoad (:4200/require.js:1717)
```

</details>

## Fix

Add a `--release` flag as well.

### What does not work

Running a debug build.

## Why this is a Flutter issue

See  https://github.com/flutter/flutter/issues/64011 for reference.

cc @sigmundch

**Comment by darshankawar** [comment]
Hi @creativecreatorormaybenot 
I tried to replicate the issue on latest beta, but don't see the exception you mentioned.

<img width="1199" alt="Screenshot 2020-09-29 at 4 04 41 PM" src="https://user-images.githubusercontent.com/67046386/94547918-933ef680-026d-11eb-90d9-1e1933f4792f.png">

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel beta, 1.22.0-12.2.pre, on Mac OS X 10.15.4 19E2269, locale
    en-IN)
    • Flutter version 1.22.0-12.2.pre at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 2bafdc8226 (4 days ago), 2020-09-25 12:48:22 -0700
    • Engine revision f763b5b9b9
    • Dart version 2.10.0 (build 2.10.0-110.5.beta)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

 
[✓] Connected device (2 available)            
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome
      85.0.4183.121

• No issues found!

```

</details>

Not sure if I missed anything from the steps you mentioned above, but is there anything else to replicate it ?
Thanks.

**Comment by creativecreatorormaybenot** [comment]
@darshankawar Sorry! I forgot to add that you need to call `mock()` in `main`, otherwise the tree shaker will remove the function.

So this is what you should have:

```dart
import 'package:flutter/material.dart';
import 'package:universal_io/io.dart';

File mock() => null;

void main() {
  mock();

  runApp(MyApp());
}
```

**Comment by creativecreatorormaybenot** [comment]
@darshankawar Sorry! I messed up the mock: it should not return `null` but construct a `File` instead: 

```dart
import 'package:flutter/material.dart';
import 'package:universal_io/io.dart';

File mock() => File('mock');

void main() {
  runApp(MyApp());
}
```

**Comment by darshankawar** [comment]
Issue replicable on latest beta (1.22.0-12.2.pre), but doesn't occur in previous version (1.22.0-12.1.pre).
Running above sample code on latest beta doesn't load the app and throws uncaught typeError in console:

<img width="1202" alt="Screenshot 2020-09-29 at 6 52 12 PM" src="https://user-images.githubusercontent.com/67046386/94564624-d016e780-0285-11eb-95ee-efd867236711.png">

While same code runs in previous version properly rendering screen with no typeError

<img width="1201" alt="Screenshot 2020-09-29 at 6 57 13 PM" src="https://user-images.githubusercontent.com/67046386/94564671-e1f88a80-0285-11eb-8d28-4aab9a188886.png">


<details>
<summary>flutter doctor -v</summary>

```
✓] Flutter (Channel beta, 1.22.0-12.2.pre, on Mac OS X 10.15.4 19E2269, locale
    en-IN)
    • Flutter version 1.22.0-12.2.pre at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 2bafdc8226 (4 days ago), 2020-09-25 12:48:22 -0700
    • Engine revision f763b5b9b9
    • Dart version 2.10.0 (build 2.10.0-110.5.beta)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

 
[✓] Connected device (2 available)            
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome
      85.0.4183.121

• No issues found!

```

</details>

**Comment by Markzipan** [comment]
Thanks for the triage, I'll take a look.

**Comment by Markzipan** [comment]
Following up here, I have a local repro in DDC, and I'm working on a fix right now.

**Comment by aleksandr-plujnik** [comment]
any updates? version 1.23 is alive, but nothing changed.

**Comment by sigmundch** [comment]
@aleksandr-plujnik fixes are approaching. There were two very closely related issues, this one and #64011

@Markzipan's reverted the fix of #64011, so the situation described in this issue should no longer occur with that reverted. That revert is approved to be cherry-picked to the stable channel, but I don't believe it made it on time to get in Dart 2.10.1, so it will probably be part of the next hotfix release.

@Markzipan also landed a fix yesterday in Dart's master for the original issue that would not regress cases like the ones on this issue. We'll request that it be included in the beta channel

**Comment by sigmundch** [comment]
Quick update: @Markzipan fix is already in Dart's dev release 2.11.0-193.0.dev and we believe it has been rolled to flutter's master channel. The beta release is still in the works, but @creativecreatorormaybenot if it is easy for you to verify that both this and #64011 are fixed in master, that would be great information to have.

Cheers!

**Comment by creativecreatorormaybenot** [comment]
@sigmundch That is so great to hear 🚀

We have temporarily ditched the dependency in order to fix this issue, however, I was able to go back to it with a bit of effort.

Anyways, I noticed that the Dart release is not yet on `master`:

```
Flutter 1.23.0-8.0.pre.227 • channel master • https://github.com/flutter/flutter.git
Framework • revision a67ba2a2ba (45 minutes ago) • 2020-10-08 23:37:34 +0530
Engine • revision 2eac514f26
Tools • Dart 2.11.0 (build 2.11.0-186.0.dev)
```

Given my little "preface", however, I will be able to verify this once the Dart release lands on Flutter `master` 👍

**Comment by creativecreatorormaybenot** [comment]
@sigmundch Actually, I figured out how to overwrite the engine version on my own and I just did that:

```
Flutter 1.23.0-8.0.pre.227 • channel master • https://github.com/flutter/flutter.git
Framework • revision a67ba2a2ba (52 minutes ago) • 2020-10-08 23:37:34 +0530
Engine • revision 57d58e6b69
Tools • Dart 2.11.0 (build 2.11.0-200.0.dev)
```

With that, I can verify.. that both of the issues are now fixed in Dart SDK build `2.11.0-200.0.dev` 🎉  
Thanks a lot for your work @sigmundch @Markzipan ❤

**Comment by sigmundch** [comment]
Awesome, thanks so much for verifying!
